Configuration files
===================

Tools in HaploCoV require configuration files to be executed correctly. They will quit and halt their execution if these configuration files are not found in your installation. In general configuration files need to be in the **same folder** from where the utilities are executed.
While all the tools in HaploCoV that use configuration files will attempt to download the most recent version directly from this Github repository, users are kindly asked to *double check* that all the configuration files are in place before executing their analyses.

The main configuration files required by HaploCoV are:

1. *globalAnnot.gz*: a file with functional annotation of the complete collection of SARS-CoV-2 variants. This file can be generated by using CorGAT (see Chiara et al 2020), for your convenience, and up to date copy is  included in the this repository. The file is updated on a weekly basis (every Wednesday).  The most recent copy is automatically downloaded by LinToFeats.pl at every execution. 

2. *areaFile*: This configuration file defines macro-geographic areas as described in *Chiara et al 2022*. The file can be edited (see Execution of custom analyses) to specify user defined macro-geographic regions.  

3. *linDefMut*: contains defining mutations for every Pango Lineage. Updated on a weekly basis (every Wednesday).

4. *scalingFactors.csv*: this file contains the list of features used by HaploCoV to compute its "VOCness" score. Please refer to the HaploCoV paper for more details. The file is used by "report.pl" and needs to be in the same folder from where the tool is executed. Although, see above ...

Pre-computed files
==================

Along with files that are required to configure the execution of tools, the Github repository of HaploCoV does also feature several pre-computed files that can to execute your analyses.
In particular, we provide a relatively large collection of pre-computed genomic variant, that can be used along with HaploCoV to identify novel lineages/sub-lineages of SARS-CoV-2.

These precomputed sets of alleles can broadly be categorized into 3 main classes:

1. *Highly divergent genomes.* These are genomic variants found in at least 25 "highly divergent" genomic sequences, at not-overlapping intervals of time of 60 days.  Highly divergent/variable genomes are defined as those carrying at least 6 or more genomic variants that are not characteristic to their assigned lineage according to the Pango nomenclature. Intervals are computed as non-overlapping windows of 60 days, starting from Mon 12-30-2019.  Dates are expressed in "HaploCoV format", i.e offsets from the start date 2019-12-30. For example the file: *900_960_list.txt* contains allelic variants identified in at least 25 distinct highly divergent genomes between day 900 and day 960. These files are stored under the folder: **HighVar**


2. *Country specific genomic variants.* Genomic variants reaching a frequency of 1% or higher, for at least 15 days  in a country at any time point from Mon 2019-12-30.  These files are stored under the folder: **country**. Each file is named  after the corresponding country. *Thailand_list.txt* reports the list of high frequency allelic variants observed in Thailand


3. *Increased frequency genomic variants.* Genomic variants showing an increase in their prevalence of a 1.5 fold or greater in at least one country, at different months, and starting from January 2020.  These files are stored under the folder: **HighFreq**. Each file is named according to the corresponding month: *April2021_list.txt* reports allele that increased in prevalence in April 2021 <br>

All our collections of genomic variants are updated on a two-week basis, every Wednesday. All files are downloaded automatically at every new installation of HaploCoV. Specific files of interest can be downloaded when/if needed. For example, under unix systems by using the `wget` utility. Some examples are reported below:

1. HyperVariable:

::

 wget https://raw.githubusercontent.com/matteo14c/HaploCoV/master/alleleVariantsSet/HighVar/900_960_list.txt` 

2. Country:

::
 
 wget https://raw.githubusercontent.com/matteo14c/HaploCoV/master/country/alleleVariantsSet/country/Thailand_list.txt 

3. High-freq:

::

 wget https://raw.githubusercontent.com/matteo14c/HaploCoV/HighFreq/master/alleleVariantsSet/HighFreq/April2021_list.txt



